store/kv/chunk.go

```diff
@@ -0,0 +1,235 @@
+	if !tx.isChunkingNeeded(data) {
+		return tx.KeyValueTx.Insert(ctx, table, key, data)
+	}
+
+	tx.setTotalChunks(data)
+
+	return tx.executeChunks(data, func(chunk int32, chunkData []byte) error {
+		chunkKey := key
+		if chunk == 0 {
+			// first part
+			zeroth := data.CloneWithAttributesOnly(chunkData)
+			if err := tx.KeyValueTx.Insert(ctx, table, chunkKey, zeroth); err != nil {
+				return err
+			}
+		} else {
+			// we don't care about any other attributes for this case
+			chunked := internal.NewTableData(chunkData)
+			// int is encoded as int64 by FDB client so better use int64 to avoid any confusion.
+			chunkKey = append(chunkKey, []KeyPart{chunkIdentifier, int64(chunk)}...)
+
+			if err := tx.KeyValueTx.Replace(ctx, table, chunkKey, chunked, false); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+func (tx *ChunkTx) Replace(ctx context.Context, table []byte, key Key, data *internal.TableData, isUpdate bool) error {
+	if !tx.isChunkingNeeded(data) {
+		return tx.KeyValueTx.Replace(ctx, table, key, data, isUpdate)
+	}
+
+	tx.setTotalChunks(data)
+
+	// cleanup the existing range
+	if err := tx.KeyValueTx.Delete(ctx, table, key); err != nil {
+		return err
+	}
+
+	return tx.executeChunks(data, func(chunk int32, chunkData []byte) error {
+		var chunked *internal.TableData
+		chunkKey := key
+		if chunk == 0 {
+			// first part
+			chunked = data.CloneWithAttributesOnly(chunkData)
+		} else {
+			chunked = internal.NewTableData(chunkData)
+			// int is encoded as int64 by FDB client so better use int64 to avoid any confusion.
+			chunkKey = append(chunkKey, []KeyPart{chunkIdentifier, int64(chunk)}...)
+		}
+
+		if err := tx.KeyValueTx.Replace(ctx, table, chunkKey, chunked, isUpdate); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// Read needs to return chunk iterator so that it can merge and returned merged chunk to caller.
+func (tx *ChunkTx) Read(ctx context.Context, table []byte, key Key) (Iterator, error) {
+	iterator, err := tx.KeyValueTx.Read(ctx, table, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ChunkIterator{
+		Iterator: iterator,
+	}, nil
+}
+
+func (tx *ChunkTx) ReadRange(ctx context.Context, table []byte, lKey Key, rKey Key, isSnapshot bool) (Iterator, error) {
+	iterator, err := tx.KeyValueTx.ReadRange(ctx, table, lKey, rKey, isSnapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ChunkIterator{
+		Iterator: iterator,
+	}, nil
+}
+
+type ChunkIterator struct {
+	Iterator
+
+	err error
+}
+
+func (it *ChunkIterator) Next(value *KeyValue) bool {
+	if !it.Iterator.Next(value) {
+		return false
+	}
+
+	if !value.Data.IsChunkedData() {
+		return true
+	}
+
+	var buf bytes.Buffer
+	buf.Write(value.Data.RawData)
+	hasNext := false
+	for chunk := int32(1); chunk < *value.Data.TotalChunks; chunk++ {
+		var chunked KeyValue
+		if hasNext = it.Iterator.Next(&chunked); !hasNext {
+			break
+		}
+		if len(chunked.Key) < 2 {
```

Comment:
move key check to the separate function.

Reply:
I think it is small check, I won't add a function call overhead in this case.

Reply:
Compiler will inline if needed.
Function name will self document and simplify understanding.

Reply:
I mean it is single place usage why do need a separate function for this?

Reply:
Just too many IFs, if we enable cyclomatic complexity linter it'll complain.
Also:
```
for ; chunk < *value.Data.TotalChunks; chunk++ {
		var chunked KeyValue
		if hasNext = it.Iterator.Next(&chunked); !hasNext {
			break
		}
 
        if !isKeyValid(chunked.Key, chunkNo) {
          return err
        }

		buf.Write(chunked.Data.RawData)
}
```
more readable.

Reply:
OK